R/pareto_smooth.R

```diff
@@ -0,0 +1,534 @@
+#'   Pareto smoothed estimates
+#' * `convergence_rate`: Relative convergence rate for Pareto smoothed estimates
+#'
+#' @examples
+#' mu <- extract_variable_matrix(example_draws(), "mu")
+#' pareto_smooth(mu)
+#'
+#' d <- as_draws_rvars(example_draws("multi_normal"))
+#' pareto_smooth(d$Sigma)
+#' @export
+pareto_smooth <- function(x, ...) UseMethod("pareto_smooth")
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.rvar <- function(x, return_k = TRUE, extra_diags = FALSE, ...) {
+
+  if (extra_diags) {
+    return_k <- TRUE
+  }
+
+  draws_diags <- summarise_rvar_by_element_with_chains(x, pareto_smooth.default, return_k = return_k, extra_diags = extra_diags, ...)
+  dim(draws_diags) <- dim(draws_diags) %||% length(draws_diags)
+  margins <- seq_along(dim(draws_diags))
+
+  if (return_k) {
+    if (extra_diags) {
+
+      diags <- list(
+        khat = apply(draws_diags, margins, function(x) x[[1]]$diagnostics$khat),
+        min_ss = apply(draws_diags, margins, function(x) x[[1]]$diagnostics$min_ss),
+        khat_threshold = apply(draws_diags, margins, function(x) x[[1]]$diagnostics$khat_threshold),
+        convergence_rate = apply(draws_diags, margins, function(x) x[[1]]$diagnostics$convergence_rate)
+      )
+    } else {
+      diags <- list(
+        khat = apply(draws_diags, margins, function(x) x[[1]]$diagnostics$khat)
+      )
+    }
+    out <- list(
+    x = rvar(apply(draws_diags, margins, function(x) x[[1]]$x), nchains = nchains(x)),
+    diagnostics = diags
+    )
+  } else {
+    out <- rvar(apply(draws_diags, margins, function(x) x[[1]]), nchains = nchains(x))
+  }
+  out
+}
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.default <- function(x,
+                                  tail = c("both", "right", "left"),
+                                  r_eff = NULL,
+                                  ndraws_tail = NULL,
+                                  return_k = TRUE,
+                                  extra_diags = FALSE,
+                                  verbose = FALSE,
+                                  ...) {
+
+  checkmate::assert_number(ndraws_tail, null.ok = TRUE)
+  checkmate::assert_number(r_eff, null.ok = TRUE)
+  checkmate::assert_logical(extra_diags)
+  checkmate::assert_logical(return_k)
+  checkmate::assert_logical(verbose)
+
+  # check for infinite or na values
+  if (should_return_NA(x)) {
+    warning_no_call("Input contains infinite or NA values, Pareto smoothing not performed.")
+    return(list(x = x, diagnostics = NA_real_))
+  }
+
+  tail <- match.arg(tail)
+  S <- length(x)
+
+  # automatically calculate relative efficiency
+  if (is.null(r_eff)) {
+    r_eff <- ess_basic(x) / S
+  }
+
+  # automatically calculate tail length
+  if (is.null(ndraws_tail)) {
+    ndraws_tail <- ps_tail_length(S, r_eff)
+  }
+
+  if (tail == "both") {
+
+    if (ndraws_tail > S / 2) {
+      warning("Number of tail draws cannot be more than half ",
+              "the length of the draws if both tails are fit, ",
```

Comment:
the total number of draws if bot tails are fit,

Reply:
Changed this

Reply:
Also changed to ess_tail